Network/Socket/ReadShow.hs

```diff
@@ -19,77 +25,76 @@ eqFst x = \(x',_) -> x' == x
-backward :: (Eq b) => Bijection a b -> b -> a
-backward Bijection{..} = lookBackward defBwd pairs
-
--- | show function for Int-like types that encodes negative numbers
--- with leading '_' instead of '-'
-_showInt :: (Show a, Num a, Ord a) => a -> String
-_showInt n | n < 0 = let ('-':s) = show n in '_':s
-           | otherwise = show n
-
--- | parse function for Int-like types that interprets leading '_'
---   as if it were '-' instead
-_readInt :: (Read a) => String -> a
-_readInt ('_':s) = read $ '-':s
-_readInt s = read s
-
-
--- | parse a quote-separated pair into a tuple of Int-like values
---   should not be used if either type might have
---   literal quote-characters in the Read pre-image
-_parse :: (Read a, Read b) => String -> (a, b)
-_parse xy =
-  let (xs, '\'':ys) = break (=='\'') xy
-   in (_readInt xs, _readInt ys)
-{-# INLINE _parse #-}
 
--- | inverse function to _parse
---   show a tuple of Int-like values as quote-separated strings
-_show :: (Show a, Num a, Ord a,  Show b, Num b, Ord b) => (a, b) -> String
-_show (x, y) = _showInt x ++ "'" ++ _showInt y
-{-# INLINE _show #-}
+-- | Unified automorphic involution over @Either a b@ that converts between
+--   LHS and RHS elements of a list of @Pair a b@ mappings and is the identity
+--   function if no matching pair is found
+lookBetween :: (Eq a, Eq b) => [Pair a b] -> Either a b -> Either a b
+lookBetween ps = \case
+  Left  x | (_,y):_ <- filter (eqFst x) ps -> Right y
+  Right y | (x,_):_ <- filter (eqSnd y) ps -> Left  x
+  z -> z
+
+type Bijection a b = [Pair a b]
 
-defShow :: Eq a => String -> (a -> b) -> (b -> String) -> (a -> String)
-defShow name unwrap sho = \x -> name ++ (sho . unwrap $ x)
+defShow :: Eq a => String -> (a -> b) -> (Int -> b -> ShowS) -> (Int -> a -> ShowS)
+defShow name unwrap shoPrec = \i x ->
+    showParen (i > app_prec) $ ((name ++ " ") ++) . (shoPrec (i+1) . unwrap $ x)
 {-# INLINE defShow #-}
 
-defRead :: Read a => String -> (b -> a) -> (String -> b) -> (String -> a)
-defRead name wrap red = \s ->
-  case splitAt (length name) s of
-    (x, sn) | x == name -> wrap $ red sn
-    _ -> error $ "defRead: unable to parse " ++ show s
+defRead :: Eq a => String -> (b -> a) -> P.ReadPrec b -> P.ReadPrec a
+defRead name wrap redPrec = do
+    P.lift $ P.expect $ P.Ident name
+    wrap <$> P.step redPrec
 {-# INLINE defRead #-}
 
--- | Apply a precedence-invariant one-token parse function within ReadPrec monad
-tokenize :: (String -> a) -> P.ReadPrec a
-tokenize f = P.lexP >>= \(P.Ident x) -> return $ f x
-{-# INLINE tokenize #-}
+
+-- | consume and return two integer-like values from two consecutive lexical tokens
+_parse :: (Bounded a, Integral a, Bounded b, Integral b) => P.ReadPrec (a, b)
+_parse = do
+  x <- safeInt
+  y <- safeInt
+  return (x, y)
+{-# INLINE _parse #-}
+
+-- | Show value with precedence equal to app_prec (=10)
+_showInt :: (Show a) => Int -> a -> ShowS
+_showInt _ x = showsPrec app_prec x
```

Comment:
With the right (`app_prec+1`) precedence passed in by the outer `Foo a` call, we should just use the first argument here, rather than force `app_prec`.
```suggestion
_showInt i x = showsPrec i x
```
at which point `_showInt` likely becomes redundant, it is just equal to `showsPrec`.